### ISSUE TITLE
Intermittent 403 errors when calling the api in SharePoint Online

### ISSUE THREAD

**Issue by nero120** [SharePoint/sp-dev-docs#1746]
#### Category
- [x] Question
- [ ] Typo
- [ ] Bug
- [ ] Additional article idea

Over the past few weeks we've been seeing an intermittent issue where users are getting 403 errors when accessing the api. This is a hard one to investigate as it doesn't happen to everyone, but it's happening to enough that issues have been raised by separate users.

Our SPFX webpart is using React and has been added to a "modern" page in our SharePoint Online site. During the `componentDidMount` event it calls the api (using [@pnp/sp](https://pnp.github.io/pnpjs/sp/index.html)) multiple times which is where the issue is occurring. Specifically, it seems that when the web part attempts to call `/_api/web/getUserEffectivePermissions` these users are receiving a 403 access denied, with the following response body:

```
<m:error xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata">
  <m:code>-2147024891, System.UnauthorizedAccessException</m:code>
  <m:message xml:lang="en-US">
    Access denied. You do not have permission to perform this action or access this resource.
  </m:message>
</m:error>
```

The users have been advised to hard refresh (ctrl + F5) the page when this issue occurs but this apparently does not make any difference. I suspected that the issue was that the user's session tokens were expiring which was causing the error, but if this is the case refreshing the page should re-authenticate them if their session has expired.

One user contacted me when this issue occurred, and since they were using chrome I had them trying refreshing the page and opening the failing request in a new tab (via dev tools), and they sent me the following:

![image](https://user-images.githubusercontent.com/4236238/39198848-001196aa-47e0-11e8-9ed9-57300163ab24.png)

I asked them to sign out of the site and sign back in, which finally did resolve the issue.

This is is particularly confusing as the users should not be getting 403s when they are authenticated, and if their session has expired then refreshing the page should resolve this. Has there been some change rolled out recently which could account for this issue, or is there any other explanation that I'm missing?

**Comment by patmill** [comment]
This isn't specifically an answer to the Access Denied error, but you may be able to skip the whole code path (and be faster), by calling this.context.pageContext.web.permissions, assuming that the web in question is the current web.

**Comment by semopz** [comment]
I can see the same thing on a list object. I'm using @pnp/sp 2.0.0 and querying a list on a remote web supplied in baseUrl.

Are there any known workarounds considering that the list might have unique permissions and/or be on a different SPWeb to where the SPFx solution is running on?

**Comment by semopz** [comment]
To answer my own question, /EffectiveBasePermissions seems to work pretty well if all you're interested in is the current user's permissions.

**Comment by gregwilton** [comment]
I ran into the same issue using the `getCurrentUserEffectivePermissions`
function to test permissions on a list (using `@pnp/sp@2.0.2`).

I was also able to resolve by using `effectiveBasePermissions` instead but
during my testing I found that granting users the "Enumerate permissions"
permission level also resolved the issue. My understanding is that "Enumerate
permissions" should only be required when querying permissions for other users
so it's odd that it was needed to query the current user.

The underlying `getUserEffectivePermissions(@user)` request URL was identical
for cases where it failed with a 403 Forbidden and when it succeeded. It seems
like sometimes SharePoint just wasn't recognising the `user` value as the current
user.
